AsyncFunc with docs

eulerfx opened the pull request.

eulerfx committed:
Async.bindChoice and Choice helpers

dsyme commented on docs/content/library/AsyncArrow.fsx:
```diff
@@ -0,0 +1,339 @@
+# F# Async: AsyncArrow
+
+An `AsyncArrow<'a, 'b>` is a function which takes a value of type `'a` as input and returns
+and async computation which produces a value of type `'b`. An async arrow is represented
```
typo "and"

dsyme commented on docs/content/library/AsyncArrow.fsx:
```diff
@@ -0,0 +1,339 @@
+(**
+
+# F# Async: AsyncArrow
```
I think I much prefer the name `AsyncFunc`.  Is there any reason not to use that?  I think of it as almost standard F# terminology - "an asynchronous function is a function `T -> Async<U>`"

eulerfx replied:
No particular reason, I'll change it to `AsyncFunc`. The reason I went with arrow is because it was an existing term and sounded at least somewhat intuitive. I initially considered `AsyncFun`, but clearly that is a bit much.

dsyme commented on docs/content/library/AsyncArrow.fsx:
```diff
@@ -0,0 +1,339 @@
+
+
+(**
+## Motivation
```
A section could be added here on "If you know F# Async, then you've already met some asynchronous functions" and then give examples.

eulerfx replied:
Ok, will do.

dsyme commented on docs/content/library/AsyncArrow.fsx:
```diff
@@ -0,0 +1,339 @@
+(**
+The value `httpTimer` is a function which takes an HTTP service and produces another HTTP service which wraps the
+argument service with a timer. If we view the HTTP service as an async arrow `AsyncArrow<HttpReq, HttpRes>`
+then the above value is a mapping between async arrows - an async filter of type `AsyncFilter<HttpReq, HttpRes>`.
```
AsyncFilter is a strange name to use - it's not a filter as in "List.filter".  I don't even know if this concept needs a name - it's just a functional combinator like `(T -> U) -> (T ->U)` is a functional combinator.

eulerfx replied:
I borrowed the name from the paper linked at the bottom _Your server as a function_. The name makes sense in the context of that paper where they use it in the same way an ASP.NET MVC action filter is used. This is also the same way it is used in the example. I do agree however that the name is confusing, but the concept as a whole seems useful nevertheless.

EDIT:

The reason I think the concept is useful is so that you can refer to a timing filter, a caching filter, a logging filter, etc. The name isn't ideal, any thoughts on an alternative?

eulerfx commented:
I renamed `AsyncArrow` to `AsyncFunc`. I also updated the wording and code samples in the docs, fixed some typos and updated the section at the bottom `Relationship to Async`. I don't yet have a better name for `AsyncFilter`.

eulerfx commented:
I removed `AsyncFilter` from the library but left it as an example in the docs. cc @dsyme

eulerfx commented:
@dsyme @forki Any more thoughts on this? I can close this and put into a different branch for now, to make way for some less controversial stuff.

eulerfx commented:
Closing for now. Moved changes to branch in my fork.
